app/scripts/modules/core/src/pipeline/config/stages/googleCloudBuild/GoogleCloudBuildStageForm.tsx

```diff
@@ -1,234 +1,220 @@
 import React from 'react';
-import Select, { Option } from 'react-select';
+import { Option } from 'react-select';
 import { get } from 'lodash';
 
 import {
   ArtifactTypePatterns,
+  excludeAllTypesExcept,
+  FormikFormField,
   IArtifact,
   IExpectedArtifact,
+  IFormInputProps,
   IFormikStageConfigInjectedProps,
+  IgorService,
   IPipeline,
-  PreRewriteStageArtifactSelector,
   RadioButtonInput,
-  SETTINGS,
-  StageArtifactSelector,
-  StageConfigField,
+  ReactSelectInput,
+  StageArtifactSelectorDelegate,
+  TextInput,
+  useData,
   yamlDocumentsToString,
   YamlEditor,
-  TextInput,
-  IGcbTrigger,
 } from '@spinnaker/core';
 
-export enum buildDefinitionSources {
-  ARTIFACT = 'artifact',
-  TEXT = 'text',
-  TRIGGER = 'trigger',
-}
-
-export enum triggerType {
-  BRANCH = 'branchName',
-  TAG = 'tagName',
-  COMMIT = 'commitSha',
-}
+import { BuildDefinitionSource, TriggerType } from './IGoogleCloudBuildStage';
 
 interface IGoogleCloudBuildStageFormProps {
-  googleCloudBuildAccounts: string[];
-  gcbTriggers: IGcbTrigger[];
   updatePipeline: (pipeline: IPipeline) => void;
-  fetchGcbTriggers: (account: string) => void;
 }
 
-interface IGoogleCloudBuildStageFormState {
-  rawBuildDefinitionYaml: string;
-}
-
-export class GoogleCloudBuildStageForm extends React.Component<
-  IGoogleCloudBuildStageFormProps & IFormikStageConfigInjectedProps,
-  IGoogleCloudBuildStageFormState
-> {
-  public constructor(props: IGoogleCloudBuildStageFormProps & IFormikStageConfigInjectedProps) {
-    super(props);
-    const stage = props.formik.values;
-    this.state = {
-      rawBuildDefinitionYaml: stage.buildDefinition ? yamlDocumentsToString([stage.buildDefinition]) : '',
-    };
-  }
-
-  private onYamlChange = (rawYaml: string, yamlDocs: any): void => {
-    this.setState({ rawBuildDefinitionYaml: rawYaml });
+const SOURCE_OPTIONS: Array<Option<string>> = [
+  { value: BuildDefinitionSource.TEXT, label: 'Text' },
+  { value: BuildDefinitionSource.ARTIFACT, label: 'Artifact' },
+  { value: BuildDefinitionSource.TRIGGER, label: 'Trigger' },
+];
+
+const TRIGGER_TYPE_OPTIONS: Array<Option<string>> = [
+  { value: TriggerType.BRANCH, label: 'Branch name' },
+  { value: TriggerType.TAG, label: 'Tag Name' },
+  { value: TriggerType.COMMIT, label: 'Commit SHA' },
+];
+
+const EXCLUDED_ARTIFACT_TYPES: RegExp[] = excludeAllTypesExcept(
+  ArtifactTypePatterns.BITBUCKET_FILE,
+  ArtifactTypePatterns.CUSTOM_OBJECT,
+  ArtifactTypePatterns.GCS_OBJECT,
+  ArtifactTypePatterns.GITHUB_FILE,
+  ArtifactTypePatterns.GITLAB_FILE,
+  ArtifactTypePatterns.HTTP_FILE,
+  ArtifactTypePatterns.S3_OBJECT,
+);
+
+export function GoogleCloudBuildStageForm(props: IGoogleCloudBuildStageFormProps & IFormikStageConfigInjectedProps) {
+  const stage = props.formik.values;
+
+  const [rawBuildDefinitionYaml, setRawBuildDefinitionYaml] = React.useState(
+    stage.buildDefinition ? yamlDocumentsToString([stage.buildDefinition]) : '',
```

Comment:
Is `yamlDocumentsToString` expensive computationally? This will get run on every render.

Reply:
Great call, updated to [lazily initialize the state](https://reactjs.org/docs/hooks-reference.html#lazy-initial-state) with a function instead.